lib/ui/painting/fragment_shader.cc

```diff
@@ -0,0 +1,97 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "flutter/lib/ui/painting/fragment_shader.h"
+
+#include "flutter/lib/ui/ui_dart_state.h"
+#include "third_party/skia/include/core/SkString.h"
+#include "third_party/tonic/converter/dart_converter.h"
+#include "third_party/tonic/dart_args.h"
+#include "third_party/tonic/dart_binding_macros.h"
+#include "third_party/tonic/dart_library_natives.h"
+
+
+using tonic::ToDart;
+
+namespace flutter {
+
+static void FragmentShader_constructor(Dart_NativeArguments args) {
+  DartCallConstructor(&FragmentShader::Create, args);
+}
+
+IMPLEMENT_WRAPPERTYPEINFO(ui, FragmentShader);
+
+#define FOR_EACH_BINDING(V) \
+  V(FragmentShader, initWithSource) \
+  V(FragmentShader, setTime) \
+  V(FragmentShader, setImage)
+
+FOR_EACH_BINDING(DART_NATIVE_CALLBACK)
+
+void FragmentShader::initWithSource(const std::string& source) {
+  initEffect(SkString(source.c_str()));
+  setShader();
+}
+
+void FragmentShader::setTime(float time) {
+  t_ = time;
+  setShader();
+}
+
+void FragmentShader::setImage(CanvasImage* image,
+                              SkTileMode tmx,
+                              SkTileMode tmy,
+                              const tonic::Float64List& matrix4) {
+  SkMatrix sk_matrix = ToSkMatrix(matrix4);
+  input_ = image->image()->makeShader(tmx, tmy, &sk_matrix);
+  setShader();
+}
+
+void FragmentShader::initEffect(SkString sksl) {
+  SkRuntimeEffect::EffectResult effect_result_ = SkRuntimeEffect::Make(sksl);
```

Comment:
style: no need for the underscore since it's not a class member

this also may be more readable if you use c++'s `std::tie`

```
SkString err;
std::tie(runtime_effect_, err) = SkRuntimeEffect::Make(sksl);
```